tfx/experimental/distributed_inference/graphdef_experiments/subgraph_partitioning/graph_partition.py

```diff
@@ -0,0 +1,518 @@
+
+      else:
+        subgraph.node.append(current_node)
+        queue.extend(node_name_to_node_def[current_node_name].input)
+
+  return ExecutionSpec(
+      subgraph=subgraph,
+      input_names=_get_input_names(subgraph),
+      output_names=output_node_names,
+      is_remote_op=False)
+
+
+def _create_placeholder_node_from_existing_node(
+    node: node_def_pb2.NodeDef, graph: tf.Graph) -> node_def_pb2.NodeDef:
+  """Creates a placeholder node to represent an existing node.
+
+  Some partitioned subgraphs may require inputs that are loaded or computed
+  previously. Hence, we replace the input nodes with placeholder nodes that
+  share the same name, shape, and dtype. Now the inputs become placeholders
+  inside partitioned subgraphs, and can be loaded by feed dicts at the runtime.
+
+  Args:
+    node: A `NodeDef` proto for the existing node.
+    graph: A tf.Graph instance for the graph that contains the existing node.
+
+  Returns:
+    A `NodeDef` proto that stores a placeholder node.
+  """
+  operation = graph.get_operation_by_name('import/%s' % (node.name))
+  output_tensor = operation.outputs[0]
+
+  with tf.compat.v1.Session(graph=tf.Graph()) as sess:
+    tf.compat.v1.placeholder(dtype=output_tensor.dtype,
+                             shape=output_tensor.shape,
+                             name=node.name)
+    return sess.graph_def.node[0]
+
+
+def _get_input_names(subgraph: graph_pb2.GraphDef) -> List[Text]:
+  input_names = [node.name for node in subgraph.node
+                 if _is_placeholder_op(node)]
+  return input_names
+
+
+def _get_non_input_names(subgraph: graph_pb2.GraphDef) -> List[Text]:
+  non_input_names = [node.name for node in subgraph.node
+                     if not _is_placeholder_op(node)]
+  return non_input_names
+
+
+def _partition_one_remote_op_layer(
+    remote_op_layer: List[Text],
+    node_name_to_node_def: Mapping[Text, node_def_pb2.NodeDef]
+) -> List[ExecutionSpec]:
+  """Constructs ExecutionSpecs for a remote op layer.
+
+  As discussed in _get_execution_specs(), a remote op layer contains one
+  or more remote ops having no dependencies on each other. However, instead of
+  having one ExecutionSpec to store a layer (as it is with subgraph layer),
+  we use multiple ExecutionSpecs to represent a remote op layer.
+
+  Args:
+    remote_op_layer: A list of remote op names for a remote op layer.
+    node_name_to_node_def: A mapping from node names to `NodeDef` protos.
+
+  Returns:
+    A list of ExecutionSpecs representing a remote op layer.
+  """
+  list_of_specs = []
+
+  for remote_op_name in remote_op_layer:
+    spec = ExecutionSpec(
+        subgraph=None,
+        input_names=list(node_name_to_node_def[remote_op_name].input),
+        output_names=[remote_op_name],
+        is_remote_op=True)
+    list_of_specs.append(spec)
+
+  return list_of_specs
+
+
+def _modify_execution_specs(specs: List[ExecutionSpec]) -> None:
+  """Modifies the execution specs.
```

Comment:
Name this something more descriptive (modify to do what?)

Reply:
Addressed. _modify_execution_specs_for_input_validity